### PR TITLE
[memcache] Split one array of memcache index to more indexes in order to better performance

### PR DESCRIPTION
**Pull request by csthomas**
#### Purpose

I want to fix a few bugs in memcache structure:
- stop memcache extension to always serialize/unserialize whole cache keys in one $index var
- split one locked index to more separated indexes (each per group_key)
- change index keys from auto increment long integer to keys which are group_keys. After a long running time auto increment key may exceed max long integer limit. 
#### Summary of Changes

**Index structure before:**

``` php
array(
    stdobject(name="[secret]-cache-[com-content]-[cache_id]", size=1024),
    stdobject(name="[secret]-cache-[com-content]-[cache_id]", size=1024),
    stdobject(name="[secret]-cache-[com-content]-[cache_id]", size=1024),
    stdobject(name="[secret]-cache-[com-content]-[cache_id]", size=1024),
    ...
)
```

**Index structure after:**
Now index is separated into more arrays ($index and $index2 per group key)

$index is array of group_keys. Example: array("com_contentE", "com_contentF");

``` php
$index = array(
    [_system],
    [com_contentA],
    [com_contentB],
    [com_contentC],
    ...,
    [com_contentE],
    [com_contentF],
    ...
);

// index2 has address at "[secret]-[group_key]-index]"
$index2 = array(
    "[com_content]-[md5sum_of_cache]" => 1025, 
    "[com_content]-[md5sum_of_cache]" => 1325, 
    "[com_content]-[md5sum_of_cache]" => 1425, 
    ...
);
```

Instead of 1 index we will have 1 + total group_keys indexes and then lock time won't be a problem.

This patch contains  some debug staff.

You can see log debug in logs/memcache.php after you click on "Clear Expired Cache" on backend.
#### Test instruction

In progress. Please code review before.

### CODE REVIEW

**Comment by csthomas** [comment]
Current version should also work with prefix platform. 
Please test.

My English is not so good. I have a little problem with explain every details.
But I can explain If someone ask for certain line / method.

**Comment by csthomas** [comment]
I'm sure that my patch can also be ported for memcached.
After that a problem described on http://forum.joomla.org/viewtopic.php?f=709&t=924389
will be fixed.

Memcache(-d) version should be as fast as file cache or faster but practise show that file cache is faster.

File: libraries/joomla/cache/storage/memcache.php [hunk 1]
```diff
@@ -295,19 +501,19 @@ public function clean($group, $mode = null)
 	 */
 	public static function isSupported()
 	{
-		// First check if the PHP requirements are met
```
Comment (mbabker) [inline libraries/joomla/cache/storage/memcache.php]: The changes in this method should be reverted.  We very explicitly changed the `isSupported` method to test whether the Memcache instance can be connected to via the current configuration instead of only checking whether the PHP extension is enabled.

File: libraries/joomla/cache/storage/memcache.php [hunk 2]
```diff
@@ -61,35 +61,79 @@ public function __construct($options = array())
...
+		}
+		else
+		{
+			$result = @static::$_db->connect($host, $port);
+		}
 
-		if ($memcachetest == false)
+		if (!$result)
 		{
-			throw new RuntimeException('Could not connect to memcache server', 404);
```
Comment (mbabker) [inline libraries/joomla/cache/storage/memcache.php]: This change should be reverted.  A connection fail should cause an Exception to be thrown.  JError is deprecated and new uses should not be added, especially with different behavior for front/backend.

File: libraries/joomla/cache/storage/memcache.php [hunk 3]
```diff
@@ -61,35 +61,79 @@ public function __construct($options = array())
...
+			{
+				JError::raiseWarning(500, 'Memcache connection failed');
+			}
+
+			return false;
 		}
 
-		// Memcahed has no list keys, we do our own accounting, initialise key index
-		if (static::$_db->get($this->_hash . '-index') === false)
+		jimport('joomla.log.log');
```
Comment (mbabker) [inline libraries/joomla/cache/storage/memcache.php]: If you're going to leave your debugging stuff in here, it needs to be cleaned up.  This `jimport()` is unnecessary as the class is autoloaded.  You only need it for classes that don't meet the autoloader requirements.

File: libraries/joomla/cache/storage/memcache.php [hunk 4]
```diff
@@ -157,6 +216,28 @@ public function getAll()
...
+	 * Get group room for cache key
+	 *
+	 * @param   string  $group         The cache data group
+	 * @param   string  $cache_id_sfx  The cache id suffix
+
+	 * @return  string  Group room
+	 *
+	 * @since   3.6
+	 */
+	public static function getGroupRoom($group, $cache_id_sfx)
```
Comment (mbabker) [inline libraries/joomla/cache/storage/memcache.php]: Can you use a more descriptive method name or explain in the method's doc block what exactly "group room" is?

**Comment by csthomas** [comment]
I will comment more later. Currently I remove deprecated stuff.

For current version of joomla I can not agree that I should revert to:
**throw new RuntimeException('Could not connect to memcache server');**
so I use:
**$app->enqueueMessage('Could not connect to memcache server', 'error');**

When I reverted to throw exception and memcached server failed then I could not login to backend.
With enqueueMessage I can and I see only error.
(The main result about success connection is boolean value from return)

I think that when memcached server failed then joomla should still working. 
**This is not major part of joomla that shut down all backend or frontend.**

Admin should have possibility to login and fix the problem in configuration  or restart memcached server after he see error message. 
Joomla will be only slower for time when memcached server is down.

__About isSupport() method__

When admin want to use memcached from not standard address (ex. other machine, unix socket, other port)  then on joomla <= 3.5.1 admin can not see field to change host and port on global config.
This is silly.

I know that there is a workaround. Admin has to start memcached server on standard local address (localhost:11211) and then change to some other address, but this is not a solution.

**Comment by mbabker** [comment]
As a high level API class, the cache storage adapters shouldn't be tied to
the application classes or making decisions on whether a message gets
enqueued for display or not.  The adapter should simply fail and a lower
level class should be handling that error.  So yes, the adapter should be
throwing an Exception and only that.  Tying it to the application classes
creates a very deep level of coupling to those specific implementations and
the static global factory, both of which should be minimized as practical
in the high level library classes.

**Comment by csthomas** [comment]
You say about encapsulation (and I may agree and revert to throw exception) but what should I do if outer function that should catch exception and take decision does not do it.

Maybe can I write it as a temporary solution and add TODO comment.
Redis works in that way.

Other way is to redesign libraries/joomla/cache/cache.php

```php
// Get the storage
		$handler = $this->_getStorage();

		if (!($handler instanceof Exception) && $this->_options['caching'])
		{
			return $handler->get($id, $group, $this->_options['checkTime']);
		}

		return false;
```

and replace $handler isinstanceof Exception by try catch.
I do not found situation where $handler is a exception, but getStorege() can throw exception which is not catched.

File: libraries/joomla/cache/storage/memcache.php (hunk 2 above)
Comment (csthomas) [inline libraries/joomla/cache/storage/memcache.php]: Done.
I have written $app->enqueuMessage() in libraries/joomla/cache/cache.php.
If there also is not an acceptable  place then I will have a little problem where it should be. 
May be in JFactory::getCache() but it probably too far.

File: libraries/joomla/cache/storage/memcache.php (hunk 1 above)
Comment (csthomas) [inline libraries/joomla/cache/storage/memcache.php]: I see that from the other side. 

When I write that pull I look at other storages, the last time in redis.php.

Method isSupport() should only say that the storage is available and will be working if you know host and port or password or some other credentials to external cache.

You want to treat isSupport() as isConnected().
Lots of storages does not need to use connect because it is in some way built-in in php, windows, etc.

There is two group of storages:
1) which use built-in cache, no need additional access, host, port, etc
2) external cache that has default address, but may have other address and should distinguish isSupported and isConnected.

In short: I want to know when my php at least support pecl memcache/memcached/redis and this is something else when joomla say me that I have a working connection.

I think about adding to storage.php one method:
```php
	/**
	 * Test to see if the storage connection is available.
	 *
	 * @return  boolean
	 *
	 * @since   3.6
	 */
	public static function isConnected()
	{
		return static::isSupported();
	}
```

For memcache/memcached/redit this will be overridden and test real connection.

**Comment by csthomas** [comment]
__Cache index in memcache storage of joomla 3.5__
Example cache id:
```php
$cache_id = "11112222333344445555666677778888-cache-_system-429adb9b368342c818ec393fb9780000"
```

**JOOMLA 3.5 index structure:**

```php
$index = array(
    stdClass($name='11112222333344445555666677778888-cache-_system-429adb9b368342c818ec393fb9780000', $size=1024),
    stdClass($name='11112222333344445555666677778888-cache-com_content-029adb9b368342c818ec393fb9780000', $size=1024),
    stdClass($name='11112222333344445555666677778888-cache-com_content-039adb9b368342c818ec393fb9780000', $size=1024),
    stdClass($name='11112222333344445555666677778888-cache-com_content-349adb9b368342c818ec393fb9780000', $size=1024),
    stdClass($name='11112222333344445555666677778888-cache-com_content-329adb9b368342c818ec393fb9780000', $size=1024),
    stdClass($name='11112222333344445555666677778888-cache-mod_menu-139adb9b368342c818ec393fb9780000', $size=1024),
    stdClass($name='11112222333344445555666677778888-cache-mod_menu-149adb9b368342c818ec393fb9780000', $size=1024),
    stdClass($name='11112222333344445555666677778888-cache-mod_menu-229adb9b368342c818ec393fb9780000', $size=1024),
    stdClass($name='11112222333344445555666677778888-cache-mod_menu-239adb9b368342c818ec393fb9780000', $size=1024),
    stdClass($name='11112222333344445555666677778888-cache-mod_menu-249adb9b368342c818ec393fb9780000', $size=1024),
    stdClass($name='11112222333344445555666677778888-cache-mod_menu-259adb9b368342c818ec393fb9780000', $size=1024),
    ...
    stdClass($name='11112222333344445555666677778888-cache-mod_menu-269adb9b368342c818ec393fb9780000', $size=1024),
);
```

Serialize and unserialize $index for methods like get/remove/clean/store takes a lots of time.
Array keys are numbers (long int) and may overflow for long running memcache/php.

__Cache index in the new memcache storage will be split into more indexes__

Primary $index and secondary indexes $index2 (per group key).

Group key is a combination of group name ex. 'com_content' and first char (hex value 0-9a-f) from md5sum or only group name otherwise:

11112222333344445555666677778888-cache-**mod_menu-2**69adb9b368342c818ec393fb9780000
11112222333344445555666677778888-cache-_**system**-429adb9b368342c818ec393fb9780000


For technical reason I have to replaced group key 'mod_menu2' to 'mod_menuC'.

Group key is generated by code below:
```php
		$key = $group;

		// If the group does not begin with '_' and does not end with number
		// then add suffix A or B or ... or P.
		// For groups like _system or com_custom_part1 do not add suffix
		if ($group[0] !== '_' && !ctype_digit(substr($group, -1, 1)))
		{
			// The cache id suffix looks like:
			// com_content-329adb1b3633424818ec393fb9780000 and then key will be com_contentD
			// Total can be 16 different keys per one group
			$key .= chr(hexdec($cache_id_sfx[strlen($group) + 1]) + 65);
		}

		return $key;
```

Now primary index is:

```php
$index = array(
	'_ssytem',
	'com_contentA',
	'com_contentD',
	'mod_menuB',
	'mod_menuC',
	...
);
```

and secondary $index2 (per group key) looks like this example for 'mod_menuC'

```php
//$index2 = $memcache→get('11112222333344445555666677778888-mod_menuC-index');
$index2 = array(
	'mod_menu-229adb9b368342c818ec393fb9780000' => 1024,
	'mod_menu-239adb9b368342c818ec393fb9780000' => 1024,
	'mod_menu-249adb9b368342c818ec393fb9780000' => 1024,
	'mod_menu-259adb9b368342c818ec393fb9780000' => 1024,
);
```
It more looks like a tree structure, where index is parent and indexes of index2 are leafs.

**Comment by csthomas** [comment]
I have plan to remove debug logs and memcache log file at all.
Error/warning logs will be changed to the same way as is in redis.

I only want to stay with debug logs for a testing and better understanding current changes.
If there is still issue with my description may be should I create some graph about cache indexing?

**Comment by csthomas** [comment]
I have redesign debug logs to be more readable and add a few other small fixes.

What do you think now?

**Comment by csthomas** [comment]
After testing on production server with memcache compression enabled I saw that memcache did not use compression almost at all.

Constructor of JCacheStorageMemcache run getConnection() only once per php process.
Memcache object is created only once per process and stored in static:$_db. 

Each module, component, etc. create new JCacheStorageMemcache() but only first had
compression set by $this->_compress = $config->get('memcache_compress', false).

**I decided to change** $this->_compress and $this->_persistent to static values, the same as static::$_db.


### Error logs in logs/memcache.php look like

```
Saving data failed for com_jmapA (size: 2538849, compression: 0)
```
Which means that memcached server can not store values bigger that 1MB, but when we enable compression then memcache store that variable.

```
Lock index2 failed for mod_menuK after 0.037997
```
Timeout after 0.037997 second, server is very busy

**Comment by csthomas** [comment]
As memcache is not used for php 7.x and 
I have no more interest with it I close it.
